rfd/0166-tpm-joining.md

```diff
@@ -0,0 +1,390 @@
+hashes and EKCert serials which are allowed to join.
+
+To assist in this process, a new command will be added `tbot tpm identify`:
+
+1. User runs `tbot tpm identify` on the host. This will query the EKPub and
+  EKCert from the TPM attached to the host and output these values for the user.
+2. User creates or modifies an existing `tpm` join token to include the EKPub
+  or EkCert serial of the TPM within the allow list. See the API Changes section
+  for the structure of this.
+
+### Joining
+
+The join process is roughly based on the TCG TPM 2.0 Keys for Device Identity
+and Attestation recommended "Identity Provisioning" flow. As per this document,
+the flow is designed to assure that:
+
+- The TPM is authentic (by validating the EKCert)
+- The specific TPM is authentic and is represented by the EKCert (by validating
+  that the EK in the TPM corresponds to the key in the EKCert)
+- The generated AK is authentic (resident in the same TPM as the EK)
+
+```mermaid
+sequenceDiagram
+  participant T as TPM
+  participant B as Bot
+  participant A as Auth Server
+  B->>T: Queries EK and EKCert
+  T-->>B: EK, EK handle, EKCert
+  B->>T: Request AK generation
+  T-->>B: AKPub, AK handle
+  B->>A: Starts RegisterUsingTPM RPC stream
+  B->>A: Sends EKPub, EKCert and AKPub
+  opt TPM CA configured for Join Token
+    A->>A: EKCert signature validated against CA
+  end
+  A->>A: Validates EKPub hash, EKCert serial against allow rules in Join Token 
+  A->>A: Generates Credential Activation challenge
+  A-->>B: Sends Credential Activation challenge
+  B->>T: ActivateCredential(EK Handle, AK Handle, Challenge)
+  T->>T: Solves challenge
+  T-->>B: Credential Activation solution
+  B-->>A: Credential Activation solution
+  A->>A: Validates submitted solution against the known solution
+  A->>B: Signed TLS and SSH Certificates
+```
+
+At the core of this process is the Credential Activation ceremony. This proves
+possession of the EK and AK to the Auth Server through decrypting an
+encrypted challenge. As long as the TPM is authentic (e.g the EKCert is 
+signed by a trusted manufacturer), this ceremony also proves that the EK and AK
+are resident in the same TPM and cannot be exported.
+
+### Renewal
+
+As with all delegated join methods, the renewal will simply repeat the join
+process.
+
+## API Changes
+
+### Join Token
+
+A new JoinMethod will be added - `tpm`.
+
+The existing `JoinToken` type will be extended to include additional fields for
+TPM joining:
+
+```protobuf
+// ProvisionTokenSpecV2TPM contains the TPM-specific part of the
+// ProvisionTokenSpecV2
+message ProvisionTokenSpecV2TPM {
+  message Rule {
+    // Hint is a human-readable hint for the rule. This does not impact joining
+    // but can be used for associating a business name with an EKPub or EKCert
+    // serial.
+    // Example: "build-server-100"
+    string Hint = 1  [(gogoproto.jsontag) = "hint,omitempty"];
```

Comment:
Maybe call it Name or Description?

Reply:
Yup - that's better

Reply:
d4b45864d9d6fabfc568d74f26c35ababde2105337d7af9a6605e1c56c891aa6